### PR TITLE
James 2169 : implement sharedWith field in getMailboxes

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 1]
```diff
@@ -816,9 +814,27 @@ private Rfc4314Rights myRights(MailboxSession session) throws MailboxException {
...
+    /**
+     * ACL is sensible information and as such we should expose as few information as possible
+     * to users. This method allows to filter a {@link MailboxACL} in order to present it to
+     * the connected user.
+     */
+    @VisibleForTesting static MailboxACL filteredForSession(Mailbox mailbox, MailboxACL acl, MailboxSession mailboxSession) throws UnsupportedRightException {
+        if (mailboxSession.getUser().isSameUser(mailbox.getUser())) {
+            return acl;
+        }
+        MailboxACL.EntryKey userAsKey = MailboxACL.EntryKey.createUser(mailboxSession.getUser().getUserName());
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: miss that, but createUser is a strange method name, no?

**Comment by chibenwa** [comment]
Mmmh maybe we can add the following test:
 - Alice don't have 'a' right, then can not update "sharedWith" on bob mailbox

Otherwise 👍

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        mailbox4 = testSystem.createMailbox(MailboxFixture.MAILBOX_PATH4, otherSession);
+    }
+
+    @Test
+    public void getMetadataShouldListUsersAclWhenShared() throws Exception {
+        mailboxManager.applyRightsCommand(MAILBOX_PATH1, MailboxACL.command().forUser(OTHER_USER).rights(MailboxACL.Right.Read).asAddition(), session);
+        mailboxManager.applyRightsCommand(MAILBOX_PATH1, MailboxACL.command().forUser(THIRD_USER).rights(MailboxACL.Right.Read).asAddition(), session);
+        MessageManager messageManager = mailboxManager.getMailbox(MAILBOX_PATH1, session);
+
+        MessageManager.MetaData actual = messageManager.getMetaData(false, session, MessageManager.MetaData.FetchGroup.NO_COUNT);
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageManagerTest.java]: can you name false?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+
+    @Test
+    public void getMetadataShouldNotExposeOtherUsersWhenSessionIsNotOwner() throws Exception {
+        mailboxManager.applyRightsCommand(MAILBOX_PATH1, MailboxACL.command().forUser(OTHER_USER).rights(MailboxACL.Right.Read).asAddition(), session);
+        mailboxManager.applyRightsCommand(MAILBOX_PATH1, MailboxACL.command().forUser(THIRD_USER).rights(MailboxACL.Right.Read).asAddition(), session);
+        MessageManager messageManager = mailboxManager.getMailbox(MAILBOX_PATH1, session);
+
+        MessageManager.MetaData actual = messageManager.getMetaData(false, otherSession, MessageManager.MetaData.FetchGroup.NO_COUNT);
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageManagerTest.java]: idem

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/StoreMessageManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    public void filteredForSessionShouldBeIdentityWhenOwner() throws UnsupportedRightException {
+        MailboxACL acl = new MailboxACL()
+            .apply(MailboxACL.command().rights(Right.Read, Right.Write).forUser(OTHER_USER).asAddition())
+            .apply(MailboxACL.command().rights(Right.Read, Right.Write, Right.Administer).forUser(THIRD_USER).asAddition());
+        MailboxACL actual = StoreMessageManager.filteredForSession(
+            new SimpleMailbox(MAILBOX_PATH1, UID_VALIDITY), acl, new MockMailboxSession(USER));
+        assertThat(actual).isEqualTo(acl);
+    }
+
+
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/StoreMessageManagerTest.java]: 2 lines

**Comment by mbaechler** [comment]
> Alice don't have 'a' right, then can not update "sharedWith" on bob mailbox

it's about setMailboxes and it's not in the scope

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java [hunk 5]
```diff
@@ -30,6 +30,47 @@
...
+            username = session.getUser().getUserName();
+        }
+        return new MailboxPath(namespace, username, mailbox);
+
+    }
+
+    /**
+     * Create a {@link MailboxPath} in the prive namespace of the specified user
+     */
+    public static MailboxPath forUser(String username, String fullmailboxname) {
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: It's a factory method that allow to go from a 3 strings constructor to a 2 strings factory method.
Can we agree it's already a progress ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java [hunk 6]
```diff
@@ -102,10 +103,12 @@ public void setup() throws Throwable {
 
         String domain = "domain.tld";
         alice = "alice@" + domain;
+        bob = "bob@" + domain;
         String password = "password";
         DataProbe dataProbe = jmapServer.getProbe(DataProbeImpl.class);
         dataProbe.addDomain(domain);
         dataProbe.addUser(alice, password);
+        dataProbe.addUser(bob, password);
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMailboxesMethodTest.java]: lazyness

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 1 above)
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Yes, but I already have (way) more refactoring commits than others

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageManagerTest.java (hunk 2 above)
Comment (mbaechler) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/AbstractMessageManagerTest.java]: of course

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rouazana) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: ok, let's go with it. But I still disagree on the second parameter name. If I follow the conventions in this call, it should not be a fullmailboxname.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/MailboxMetaData.java [hunk 7]
```diff
@@ -48,14 +49,9 @@
 
     public MailboxMetaData(List<MessageUid> recent, Flags permanentFlags, long uidValidity, MessageUid uidNext, long highestModSeq, long messageCount, long unseenCount, MessageUid firstUnseen, boolean writeable, boolean modSeqPermanent, MailboxACL acl) {
         super();
-        if (recent == null) {
-            this.recent = new ArrayList<>();
-        } else {
-            this.recent = recent;
-
-        }
+        this.recent = Optional.ofNullable(recent).orElseGet(ArrayList::new);
```
[the remark this answers is not in the document]
Comment (rouazana) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/MailboxMetaData.java]: add c/r

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: I modified argument name

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/MailboxMetaData.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/MailboxMetaData.java]: For this line?

I find 

```
this.recent = Optional.ofNullable(recent).orElseGet(ArrayList::new);
```

Not too long and easy to read...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: **createUserEntryKey** looks better to me.

File: mpt/impl/imap-mailbox/core/src/main/java/org/apache/james/mpt/imapmailbox/suite/AuthenticatedState.java [hunk 8]
```diff
@@ -412,14 +412,14 @@ public void testNamespaceKOREA() throws Exception {
     @Test
     public void listShouldNotListMailboxWithOtherNamspaceUS() throws Exception {
```
Comment (rouazana) [inline mpt/impl/imap-mailbox/core/src/main/java/org/apache/james/mpt/imapmailbox/suite/AuthenticatedState.java]: while you are at it s/Namspace/Namespace if this file

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java [hunk 9]
```diff
@@ -147,11 +149,14 @@ public void createMailboxShouldCreateAMailbox() throws Exception {
     }
 
     @Test
-    public void createMailboxShouldNotThrowIfMailboxAlreadyExist() throws Exception {
+    public void createMailboxShouldThrowIfMailboxAlreadyExist() throws Exception {
```
Comment (rouazana) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: s/Exist/Exists/
but why this change?

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java [hunk 10]
```diff
@@ -147,11 +149,14 @@ public void createMailboxShouldCreateAMailbox() throws Exception {
...
         MailboxPath path = MailboxPath.forUser(USER, "name");
         Mailbox mailbox = new SimpleMailbox(path, UID_VALIDITY);
         inMemoryMapperFactory.createMailboxMapper(session).save(mailbox);
-        mailboxManagerManagement.createMailbox(MailboxConstants.USER_NAMESPACE, USER, "name");
+
```
Comment (quynhn) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: Can we break the second assertion into new testing method?

**Comment by chibenwa** [comment]
CLI broken 

```
[INFO] Running org.apache.james.cli.type.CmdTypeTest
[INFO] Tests run: 45, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.139 s - in org.apache.james.cli.type.CmdTypeTest
[INFO] Running org.apache.james.cli.ServerCmdTest
Storage space allowed for Sieve scripts by default: 18 bytes
Storage space allowed for Quota Root #private&user@domain: UNLIMITED

Quota Root: #private&user@domain
Default Maximum message count Quota: 1048576

MailboxMessage count quota for #private&user@domain is: UNKNOWN / UNLIMITED


MailboxMessage count allowed for Quota Root #private&user@domain: UNLIMITED
Default Maximum Storage Quota: 1 GB

example.com exists
Storage quota for #private&user@domain is: UNKNOWN / UNLIMITED
Storage space allowed for btellier@apache.org Sieve scripts: 18 bytes
[ERROR] Tests run: 100, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.212 s <<< FAILURE! - in org.apache.james.cli.ServerCmdTest
[ERROR] createMailboxMappingCommandShouldWork(org.apache.james.cli.ServerCmdTest)  Time elapsed: 0.003 s  <<< ERROR!
java.lang.IllegalStateException: 
missing behavior definition for the preceding method call:
MailboxProbe.createMailbox("#private", "user@domain", "INBOX.test")
Usage is: expect(a.foo()).andXXX()
	at org.apache.james.cli.ServerCmdTest.createMailboxMappingCommandShouldWork(ServerCmdTest.java:314)
```

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java (hunk 9 above)
Comment (chibenwa) [reply server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: Because now the JmxProbe forward an exception (feat @mbaechler ) hence the test was failing.

**Comment by mbaechler** [comment]
I can't approve my own PR, but I would definitely do otherwise.
👍

**Comment by mbaechler** [comment]
```
[ERROR] Tests run: 9, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 14.588 s <<< FAILURE! - in org.apache.james.jmap.cassandra.access.CassandraAccessTokenRepositoryTest
[ERROR] validTokenMustBeRetrieved(org.apache.james.jmap.cassandra.access.CassandraAccessTokenRepositoryTest)  Time elapsed: 0.126 s  <<< ERROR!
java.util.concurrent.CompletionException: org.apache.james.jmap.api.access.exceptions.InvalidAccessToken: a9e17678-efe1-4288-9941-d67c3cb5afb5 is missing
Caused by: org.apache.james.jmap.api.access.exceptions.InvalidAccessToken: a9e17678-efe1-4288-9941-d67c3cb5afb5 is missing
```

**Comment by mbaechler** [comment]
test this please

**Comment by mbaechler** [comment]
```
Quota Root: #private&user@domain
Default Maximum message count Quota: 1048576

MailboxMessage count quota for #private&user@domain is: UNKNOWN / UNLIMITED


MailboxMessage count allowed for Quota Root #private&user@domain: UNLIMITED
Default Maximum Storage Quota: 1 GB

example.com exists
Storage quota for #private&user@domain is: UNKNOWN / UNLIMITED
Storage space allowed for btellier@apache.org Sieve scripts: 18 bytes
[ERROR] Tests run: 100, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.211 s <<< FAILURE! - in org.apache.james.cli.ServerCmdTest
[ERROR] createMailboxMappingCommandShouldWork(org.apache.james.cli.ServerCmdTest)  Time elapsed: 0.003 s  <<< ERROR!
java.lang.IllegalStateException: 
missing behavior definition for the preceding method call:
MailboxProbe.createMailbox("#private", "user@domain", "INBOX.test")
Usage is: expect(a.foo()).andXXX()
	at org.apache.james.cli.ServerCmdTest.createMailboxMappingCommandShouldWork(ServerCmdTest.java:314)
```
still broken at the same place then ... I thought it was fixed ... I lost my day waiting for that.

**Comment by chibenwa** [comment]
test this please

(passed in my local)

```
[ERROR] Errors: 
[ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.attachmentsShouldBeRetrievedWhenChainingSetMessagesAndGetMessagesTextAttachment:2985 » ConditionTimeout
```

**Comment by chibenwa** [comment]
Merged
